### PR TITLE
Flash_Circular_Buffer basing on Flash_map module

### PR DESCRIPTION
**Pull request by nvlsianpu**
This patch objective is addition of Flash Circular Buffer module as zephyr native code.
Code of the Flash Circular Buffer is taken from MyNewt.
API of the flash_map module is almost the same as for MyNewt. Implementation of this part is mostly reworked for zephyr, flash_area_get_sectors API and its implementation is taken form mcuboot.

Explanation for choice addition of flash_map module:
Of course the providing of substitute of flash_area primitives is intermedium step for adding other modules. I will adapt mynewt FCB (Flash Circular Buffer) to zephyr, I will take mynewt Config module (or something very similar ) as well. What I also potentially want to do providing something like mynewt’s Manufacturer meta region is – this would be useful for inform both bootloader and application about how memory map looks like. I think that zephyr flash_map will simplify support for multi-memories and multi-images mcuboot support as well or event we can support dynamic image slots allocation basing on that (so will be possible to change flash areas boundaries during FWU – we see the case for that).

Question:
Should FCB module been moved to File System directory (so Kconfig as well)?

Should I provide better way for providing user's custom flash map description instead of CONFIG_FLASH_MAP_CUSTOM option (look at https://github.com/nvlsianpu/zephyr/blob/add_fcb/subsys/storage/flash_map/CMakeLists.txt#L2 ).

### CODE REVIEW

File: subsys/flash_map/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+zephyr_sources_ifdef(CONFIG_FLASH_MAP flash_map.c)
+
+if(CONFIG_FLASH_MAP)
+zephyr_sources_ifndef(CONFIG_FLASH_MAP_CUSTOM flash_map_default.c)
```
Comment (SebastianBoe) [inline subsys/flash_map/CMakeLists.txt]: Indentation.

File: subsys/fcb/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+zephyr_sources_ifdef(CONFIG_FCB fcb_append.c  fcb.c  fcb_elem_info.c  fcb_getnext.c fcb_rotate.c  fcb_walk.c)
```
Comment (SebastianBoe) [inline subsys/fcb/CMakeLists.txt]: it is easier to read (and to diff) if each file is on it's own line.

File: subsys/flash_map/CMakeLists.txt (hunk 1 above)
Comment (SebastianBoe) [reply subsys/flash_map/CMakeLists.txt]: You have alread ifdef'd this entire directory, so you dont' need to ifdef flash_map again here. The same applies to FCB I assume.

File: include/crc8.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+#define __CRC8_H_
+
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+u8_t crc8_init(void);
+u8_t crc8_ccitt(u8_t val, void *buf, int cnt);
```
Comment (nashif) [inline include/crc8.h]: missing documentation of this API

File: drivers/crc/crc8_sw.c [hunk 4]
```diff
@@ -0,0 +1,31 @@
+/*
```
Comment (nashif) [inline drivers/crc/crc8_sw.c]: this file needs to go into lib/crc (crc has moved from drivers/)

Comment (nashif) [reply drivers/crc/crc8_sw.c]: also please add tests for this new API

File: subsys/fcb/fcb.c [hunk 5]
```diff
@@ -0,0 +1,252 @@
+/*
```
Comment (nashif) [inline subsys/fcb/fcb.c]: Dont think this makes a subsystem on its own, this belongs rather into something else, maybe subsys/storage.

**Review by nashif — CHANGES_REQUESTED** [review]

File: subsys/fcb/fcb.c (hunk 5 above)
Comment (nashif) [reply subsys/fcb/fcb.c]: or under filesystem... (fs)

File: samples/basic/blinky/prj.conf [hunk 6]
```diff
@@ -1,2 +1,3 @@
 CONFIG_GPIO=y
 CONFIG_SERIAL=n
+CONFIG_FLASH_MAP=y
```
Comment (nashif) [inline samples/basic/blinky/prj.conf]: why is this needed here?

File: subsys/flash_map/Kconfig [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+	help
+	  Enable support of flash map abstraction.
+
+config FLASH_MAP_CUSTOM
+	bool
+	prompt "Custom flash map description"
+	default n
+	depends on FLASH_MAP
+	help
+	  User must provide custom flash map description in place of default on
```
Comment (nashif) [inline subsys/flash_map/Kconfig]: rephrase this help message, something is missing.

File: subsys/flash_map/flash_map.c [hunk 8]
```diff
@@ -0,0 +1,270 @@
+/*
```
Comment (nashif) [inline subsys/flash_map/flash_map.c]: should not be in its own subsystem, this needs to be either part of fs or a new subsystem with generic name, like storage for example.

File: subsys/flash_map/flash_map.c [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+		return -1;
+	}
+
+	dev = get_flah_dev_from_id(fa->fa_device_id);
+
+	return flash_read(dev, fa->fa_off + off, dst, len);
+}
+
+int flash_area_write(const struct flash_area *fa, off_t off, const void *src,
+			 size_t len)
```
Comment (nashif) [inline subsys/flash_map/flash_map.c]: align

File: subsys/flash_map/flash_map.c [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+int flash_area_get_sectors(int idx, uint32_t *cnt, struct flash_sector *ret)
+{
+	struct layout_data data;
+
+	return flash_area_layout(idx, cnt, ret, get_sectors_cb, &data);
+}
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
+int flash_area_read(const struct flash_area *fa, off_t off, void *dst,
+			size_t len)
```
Comment (nashif) [inline subsys/flash_map/flash_map.c]: align

File: tests/subsys/fcb/testcase.yaml [hunk 11]
```diff
@@ -0,0 +1,5 @@
+tests:
+  - test:
```
Comment (nashif) [inline tests/subsys/fcb/testcase.yaml]: should not be a sequence, needs to be a map.

File: drivers/crc/crc8_sw.c [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+	int i;
+	u8_t *p = buf;
+
+	for (i = 0; i < cnt; i++) {
+		val ^= p[i];
+		val = (val << 4) ^ crc8_ccitt_small_table[val >> 4];
+		val = (val << 4) ^ crc8_ccitt_small_table[val >> 4];
+	}
+	return val;
+}
```
Comment (SebastianBoe) [inline drivers/crc/crc8_sw.c]: Shouldn't this be lib/crc/crc8_sw.c?

**Review by SebastianBoe — CHANGES_REQUESTED** [review]

**Comment by SebastianBoe** [comment]
> Should I provide better way for providing user's custom flash map description instead of CONFIG_FLASH_MAP_CUSTOM option.

Perhaps allowing Kconfig'uration of a handful of predefined 'app' areas so the user can add his own areas (when he has a simple use-case with only a handful of areas) purely from Kconfig. This might
be more maintainable for some users since they won't have to copy-and-modify default_flash_map.

I don't really understand the use-case here though, so maybe I am way off :)

```
#ifdef CONFIG_FLASH_MAP_APP_AREA_0
	{
		.fa_id = 5,
		.fa_device_id = SOC_FLASH_0_ID,
		.fa_off  = CONFIG_FLASH_AREA_APP_0_OFFSET,
		.fa_size = CONFIG_FLASH_AREA_APP_0_SIZE,
	},
#endif
#ifdef CONFIG_FLASH_MAP_APP_AREA_1
	{
		.fa_id = 5,
		.fa_device_id = SOC_FLASH_0_ID,
		.fa_off  = CONFIG_FLASH_AREA_APP_1_OFFSET,
		.fa_size = CONFIG_FLASH_AREA_APP_1_SIZE,
	},
#endif

```

File: subsys/fcb/fcb.c (hunk 5 above)
Comment (carlescufi) [reply subsys/fcb/fcb.c]: I suggest we place it under filesystem

**Comment by carlescufi** [comment]
> Should FCB module been moved to File System directory (so Kconfig as well)?
Yes, I think you should place it in `subsys/fs/fcb`

File: tests/subsys/storage/flash_mpa/testcase.yaml [hunk 13]
```diff
@@ -0,0 +1,5 @@
+tests:
+  - test:
```
Comment (nashif) [inline tests/subsys/storage/flash_mpa/testcase.yaml]: still wrong

**Review by nashif — CHANGES_REQUESTED** [review]

**Comment by nvlsianpu** [comment]
@updated

File: include/flash_map/flash_map.h [hunk 14]
```diff
@@ -0,0 +1,90 @@
+/*
```
Comment (nashif) [inline include/flash_map/flash_map.h]: can we put this in include/<subsystem>/flash_map.h
same for fcb.h

**Review by nashif — CHANGES_REQUESTED** [review]

File: subsys/storage/flash_map/flash_map.c [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+	}
+
+	ret[data->ret_idx].fs_off = info->start_offset - data->area_off;
+	ret[data->ret_idx].fs_size = info->size;
+	data->ret_idx++;
+
+	return true;
+}
+
+int flash_area_get_sectors(int idx, uint32_t *cnt, struct flash_sector *ret)
```
Comment (nvlsianpu) [inline subsys/storage/flash_map/flash_map.c]: @mbolivar Should I add Linaro to copy right list? I mean that I take this form mcuboot, and it was your code.

Comment (mbolivar) [reply subsys/storage/flash_map/flash_map.c]: I think the main thing to do is preserve any copyright notices that were in the files which you're copying content from, so if there's Linaro headers in the original file, please make sure they're reproduced here, yes.

Comment (nvlsianpu) [reply subsys/storage/flash_map/flash_map.c]: So I will keep this as is - As mcuboot file from which i take this code doesn't have copyright list.

**Comment by nvlsianpu** [comment]
@SebastianBoe Thanks from Idea for description customization. As so far I didn't see real user case too I think I will be better to improve this when something will appear (so not in this patch).

File: subsys/CMakeLists.txt [hunk 16]
```diff
@@ -10,3 +10,5 @@ add_subdirectory_ifdef(CONFIG_MCUBOOT_IMG_MANAGER  dfu)
 add_subdirectory_ifdef(CONFIG_NET_BUF              net)
 add_subdirectory_ifdef(CONFIG_USB                  usb)
 add_subdirectory(random)
+add_subdirectory(storage)
+add_subdirectory_ifdef(CONFIG_FCB  fs/fcb)
```
Comment (SebastianBoe) [inline subsys/CMakeLists.txt]: Have you kept this out of the fs directory because FCB does not depend on CONFIG_FILE_SYSTEM?

We need to find some way of getting it into fs as users will expect this code to be in there. If FCB does not require CONFIG_FILE_SYSTEM you will need to refactor away the ifdef FILE_SYSTEM in this file. PM me if I am being unclear.

Comment (nvlsianpu) [reply subsys/CMakeLists.txt]: >Have you kept this out of the fs directory because FCB does not depend on CONFIG_FILE_SYSTEM?

Yes, exactly.

Comment (nvlsianpu) [reply subsys/CMakeLists.txt]: FCB code do not depend on CONFIG_FILE_SYSTEM already.

**Review by SebastianBoe — APPROVED** [review]

File: include/fcb.h [hunk 17]
```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2017 Nordic Semiconductor ASA
+ * Copyright (c) 2015 Runtime Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef __SYS_FCB_H_
```
Comment (carlescufi) [inline include/fcb.h]: Why `__SYS_FCB_H` and not `_FCB_H_`?

File: include/fcb.h [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+/*
+ * Clears FCB passed to it
+ */
+int fcb_clear(struct fcb *fcb);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* __SYS_FLASHVAR_H_ */
```
Comment (carlescufi) [inline include/fcb.h]: wrong comment

File: subsys/fs/fcb/fcb_priv.h [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+int fcb_sector_hdr_init(struct fcb *fcb, struct flash_area *fap, u16_t id);
+int fcb_sector_hdr_read(struct fcb *fcb, struct flash_area *fap,
+			struct fcb_disk_area *fdap);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif
```
Comment (carlescufi) [inline subsys/fs/fcb/fcb_priv.h]: missing comment

**Review by carlescufi — CHANGES_REQUESTED** [review]
Really nice work, just a couple of nitpicks

**Comment by nvlsianpu** [comment]
@carlescufi updated

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@mbolivar can you take another look at this when you have time?

**Comment by mbolivar** [comment]
> @mbolivar can you take another look at this when you have time?

@carlescufi I would actually prefer to ask @mike-scott to take a look at this. He's the person in our team that is most impacted

File: subsys/storage/flash_map/flash_map.c [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+	void *ret;        /* struct flash_area* or struct flash_sector* */
+	u32_t ret_idx;
+	u32_t ret_len;
+	int status;
+};
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
+struct driver_map_entry {
+	u8_t id;
+	const char * const name;
```
Comment (grandcat) [inline subsys/storage/flash_map/flash_map.c]: When implementing multiple flash support [MCUB-75](https://runtimeco.atlassian.net/projects/MCUB/issues/MCUB-75) for mcuboot,  I went a similar approach as the `driver_map_entry`, but included the a pointer to `struct device` itself.
It helps to avoid many calls to `device_get_binding(..)`. Instead, it is called once and cached.

Could this be an option?

File: subsys/storage/flash_map/flash_map.c [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+
+	return -ENOENT;
+}
+
+void flash_area_close(const struct flash_area *fa)
+{
+	/* nothing to do for now */
+}
+
+static struct device *get_flah_dev_from_id(u8_t id)
```
Comment (grandcat) [inline subsys/storage/flash_map/flash_map.c]: Typo: `get_flah_dev_from_id` -> `get_flash_dev_from_id`

File: tests/subsys/storage/flash_mpa/src/main.c [hunk 22]
```diff
@@ -0,0 +1,103 @@
+/*
```
Comment (grandcat) [inline tests/subsys/storage/flash_mpa/src/main.c]: Typo in directory:
` tests/subsys/storage/flash_mpa` -> ` tests/subsys/storage/flash_map`

**Review by nashif — APPROVED** [review]
